src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,287 @@
+    CAmount& value_ret, std::vector<CAmount>& input_fee_vec, std::vector<CAmount>& long_term_fee_vec, CAmount& fee_ret)
+{
+    out_set.clear();
+    value_ret = 0;
+
+    if (utxo_pool.empty()) {
+        return false;
+    }
+
+    int depth = 0;
+    int remaining_tries = 100000;
+    std::vector<std::pair<bool, bool>> selection; // First bool: select the utxo at this index; Second bool: traversing exclusion branch of this utxo
+    selection.assign(utxo_pool.size(), std::pair<bool, bool>(false, false));
+    bool done = false;
+    bool backtrack = false;
+
+    // Sort the utxo_pool
+    std::sort(utxo_pool.begin(), utxo_pool.end(), descending);
+
+    // Calculate lookahead: total value of the UTXO pool minus whatever has been chosen
+    CAmount lookahead = 0;
+    for (CInputCoin utxo : utxo_pool) {
+        lookahead += utxo.txout.nValue;
+    }
+
+    // Best solution
+    CAmount curr_waste = 0;
+    std::vector<std::pair<bool, bool>> best_selection;
+    CAmount best_waste = MAX_MONEY;
+
+    // Depth First search loop for choosing the UTXOs
+    while (!done)
+    {
+        if (remaining_tries <= 0) { // Too many tries, exit
+            break;
+        } else if (value_ret > target_value + cost_of_change) { // Selected value is out of range, go back and try other branch
+            backtrack = true;
+        } else if (curr_waste > best_waste) { // Don't select things which we know will be more wasteful
+            backtrack = true;
+        } else if (value_ret >= target_value) { // Selected value is within range
+            curr_waste += (value_ret - target_value); // This is the excess value which is added to the waste for the below comparison
+            if (curr_waste <= best_waste) {
+                best_selection.assign(selection.begin(), selection.end());
+                best_waste = curr_waste;
+            }
+            curr_waste -= (value_ret - target_value); // Remove the excess value as we will be selecting different coins now
+            backtrack = true;
+        } else if (depth >= (int)utxo_pool.size()) { // Reached a leaf node, no solution here
+            backtrack = true;
+        } else if (value_ret + lookahead < target_value) { // Cannot possibly reach target with the amount remaining in the lookahead
+            if (depth == 0) { // At the first utxo, no possible selections, so exit
+                break;
+            } else {
+                backtrack = true;
+            }
+        } else { // Continue down this branch
+            // Assert that this utxo is not negative. It should never be negative, effective value calculation should have removed it
+            assert(utxo_pool.at(depth).txout.nValue >= 0);
+
+            // Remove this utxo from the lookahead utxo amount
+            lookahead -= utxo_pool.at(depth).txout.nValue;
+            // Increase waste
+            curr_waste += input_fee_vec[depth] - long_term_fee_vec[depth];
```

Comment:
Could use .at() consistently instead of .at() and []

Reply:
Done.